### PR TITLE
[ntuple] Improvements to the RField read/write API

### PR DESCRIPTION
**Pull request by jblomer**
As discussed in the last I/O meeting, the current RFieldBase API for reading and writing is flawed.  The `RFieldBase::Read()` and `RFieldBase::Append()` methods take `RFieldValue` parameters, as thin wrappers around the target/source memory location. This is a misuse of the `RFieldValue` class and does not provide any additional benefit.

This PR changes the RFieldBase read/write APIs to use void pointers instead. It applies a similar change to the underlying `RColumn` read/write API (note that `RColumn` is supposed to be an internal API. NB: it still needs to move from `ROOT::Detail` to `ROOT::Internal`).

In a follow-up PR, the following change is foreseen:
  - Make `RFieldValue` an inner class of `RFieldBase`, i.e. `RFieldBase::RValue`
  - Make `RFieldBase::Read()` and `RFieldBase::Append()` protected
  - Call the `RFieldBase` read/write API only through `RFieldBase::RValue`
 
This would prevent random memory locations from using the RField read/write API. Only memory locations that come from a certain field (e.g., created by that field) will use the RField read/write API.

### CODE REVIEW

File: tree/ntuple/v7/inc/ROOT/RColumn.hxx
Comment (jalopezg-git) [inline tree/ntuple/v7/inc/ROOT/RColumn.hxx:132]: Minor suggestion: also in line 162: I would stick either to `memcpy()` or `std::memcpy()` throughout the translation unit.

**Review by jalopezg-git — COMMENTED** [review]
Good work :slightly_smiling_face:!  Here's a first round of comments!

In the current state, I see it somewhat, inconsistent, e.g. `GenerateValue()` / `DestroyValue()` return (take, respectively) a `RFieldValue`, but to read (or write) something, I need to pass a raw pointer as in
```c++
field->Read(0, value.GetRawPtr());
```

I understand that this is to be dealt with in the follow-up PR, but I wonder whether it should actually be part of the same pull request.

@Nowakus: if this change is merged, you will most likely need to slightly adjust your code.
@enirolf: this change will require rebasing/adjusting https://github.com/root-project/root/pull/12948 accordingly

**Comment by jblomer** [comment]
> @Nowakus: if this change is merged, you will most likely need to slightly adjust your code. 

@jalopezg-git can you explain? Wouldn't framework code use the `RNTupleReader` and `RNTupleWriter`?

File: tree/ntuple/v7/src/RField.cxx
Comment (jblomer) [reply tree/ntuple/v7/src/RField.cxx:210]: I'd be happy for a suggestion.  I played a bit with several `const_cast`s and considered some template mechanisms. Eventually, the most readable way seemed to copy that code. But that's certainly not ideal either.

**Comment by jalopezg-git** [comment]
> > @Nowakus: if this change is merged, you will most likely need to slightly adjust your code.
> 
> @jalopezg-git can you explain? Wouldn't framework code use the `RNTupleReader` and `RNTupleWriter`?

Oh, apologies, I guess you are right!  @Nowakus, please disregard my previous comment!

File: tree/ntuple/v7/src/RField.cxx
Comment (jalopezg-git) [reply tree/ntuple/v7/src/RField.cxx:210]: I think that move-constructing a `std::tuple` from the non-`const` qualified types should work.  Let's try with
```suggestion
std::tuple<const void **, const std::int32_t *, const std::int32_t *> GetRVecDataMembers(const void *rvecPtr)
{
   return {GetRVecDataMembers(const_cast<void *>(rvecPtr))};
}

```

Comment (jblomer) [reply tree/ntuple/v7/src/RField.cxx:210]: Right, that was my first though, too.  But unfortunately the conversion

    ‘tuple<void**, int*, int*>’ to ‘tuple<const void**, const int*, const int*>’

does not work.

Comment (jblomer) [reply tree/ntuple/v7/src/RField.cxx]: Thanks! Yep, that works.

File: tree/ntuple/v7/inc/ROOT/RColumnElement.hxx
Comment (jalopezg-git) [inline tree/ntuple/v7/inc/ROOT/RColumnElement.hxx:259]: Given that `Pack()` or `Unpack()` do not take a `RPage` as argument, IMO, the suggestion below would be more precise.
```suggestion
These specialized child classes are responsible for overriding `Pack()` / `Unpack()` for packing / unpacking elements
as appropriate.
```

**Review by jalopezg-git — APPROVED** [review]
LGTM :+1: (modulo the `GenerateValue()` / `DestroyValue()` issues already in discussion)!

Also, don't forget to apply the suggestion before merging!

**Pull request merged**
